Various minor enhancements for the 'console' command

sjg20 opened the pull request:
Provides an internal console and a way to automatically select a final state

The internal console avoids the problem of losing output when labgrid closes the serial port and re-opens it in the switch from to microcom.

The final state can be used to ensure that the board is off after use, without needing to run labgrid-client again. It is similar to the initial state in that sense.

Together, this PR allows 'labgrid client' to function as a console device suitable for use with U-Boot's test.py, for example

See here for how this is used in a real lab:

https://ci.u-boot.org/u-boot/u-boot/-/jobs/19119

**Checklist**
- [] Documentation for the feature
- [ ] Tests for the feature 
- [x] Add a section on how to use the feature to doc/usage.rst
- [x] PR has been tested
- [x] Man pages have been regenerated

sjg20 committed:
remote/client: Provide an option to acquire a place

It is common to acquire a place and then use it to access the console.
Add a -a option to support this, to avoid running labgrid-client twice,
which takes longer and makes error handling harder.

If the place is acquired, it is released on exit.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
remote/client: Using logging for progress output

Some messages are printed regardless of the logging level. Convert them
to logging statements so they can be suppressed.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
remote/client: Show the result of a failed command

When a command fails in labgrid-client it just shows the return code.
This makes it very hard to see what actually went wrong. Handle this
exception explicitly, showing the output of the tool which failed. This
allows easy debugging of the problem.

This is particularly important with things like USB loaders.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
remote/client: Use logging when acquiring / releasing

Use logging rather than print so that this output can be controlled.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
remote/client: Allow releasing a target silently

When a target may or may not be acquired, it is useful to be able to
release it without checking first. Add an option for this, so that it
is not necessary to check first.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
remote/client: Flush the console device only when needed

At present if a console device is used by a strategy driver it is then
removed by the client immediately afterwards. If the 'console' command
is used, this means that any console data is lost.

This is done so that the console device is release for microcom, so
move the logic in with the microcom logic. With the forthcoming internal
terminal, it will not be needed.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
remote/client: Move initial-state code into a function

Put this code into a function to reduce the size of the _get_target()
function.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
remote/client: Provide a way to set the strategy end-state

It is possible to set the initial state of a strategy, but not the final
state. In many cases we want to power the board off and otherwise clean
things up at the end. Provide a -e option to accomplish this.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
remote/client: Provide an internal console

At present Labgrid uses microcom as its console. This has some
limitations:

- console output is lost between when the board is reset and microcom
  connects
- txdelay cannot be handled in microcom, meaning that boards may fail
  to receive expected output
- the console may echo a few characters back to the caller in the time
  between when 'labgrid-client console' is executed and when microcom
  starts (which causes failures with U-Boot test system)

For many use cases, microcom is more than is needed, so provide a simple
internal terminal which resolved the above problems.

It is enabled by a '-i' option to the 'console' command, as well as an
environment variable, so that it can be adjustly without updating a lot
of scripts.

To exit, press Ctrl-] twice, quickly.

Series-changes: 4
- Get internal console working with qemu
- Show a prompt when starting, to indicate it is waiting for the board

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
remote/client: Move terminal handling into a separate file

There is quite a lot of code here, so move the terminal function into
its own file. This will make it easier to extend it later.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
remote/client: Split _check_allowed() into two methods

It is useful to be able to see if a place is allowed without raising an
error and needing a try...except block. Add a function to handle this
and update _check_allowed() to use it.

This will be used by the updated terminal suport.

Signed-off-by: Simon Glass <sjg@chromium.org>

aparcar commented:
Thanks for these patches, I applied most of them to my staging tree for testing in my lab. Some patches no longer apply against the master branch, could you kindly rebase this?

Regarding the internal console, does it make sense to make it more configurable so it's used for i.e. `pytest` runs, too?

sjg20 commented:
> Thanks for these patches, I applied most of them to my staging tree for testing in my lab. Some patches no longer apply against the master branch, could you kindly rebase this?

Yes, done. Thanks for looking at it.

> 
> Regarding the internal console, does it make sense to make it more configurable so it's used for i.e. `pytest` runs, too?

There is a LG_CONSOLE variable which I set in the pytest setup. Do you have any suggestions on how to make this work for you?

sjg20 commented on labgrid/remote/client.py:
```diff
@@ -988,7 +988,9 @@ async def _console(self, place, target, timeout, *, logfile=None, loop=False, li
 
-        print(f"connecting to {resource} calling {' '.join(call)}")
+        if logfile:
+            call.append(f"--logfile={logfile}")
```
OK done, thanks

sjg20 committed:
tests: Add test coverage for console and terminal handling

Add unit tests for the terminal handling in term.py, covering the
external console launch (microcom and telnet fallback), the internal
console read/write loop, and terminal setup and teardown. Also add
tests for the new client methods including is_allowed(),
set_initial_state(), set_end_state() and the new parser arguments.

Use real pipes for stdin rather than mocking os.read(), giving more
realistic coverage of the keystroke and txdelay paths. The exit-char
deadline test uses a threading.Event to synchronise with the run
loop, avoiding brittle fixed sleeps.

Fix an UnboundLocalError in term.internal() where log_fd could be
referenced in the finally block before being assigned.

Also fix a duplicate --logfile append in term.external() when using
microcom.

Signed-off-by: Simon Glass <sjg@chromium.org>
Co-developed-by: Claude Opus 4.6 <noreply@anthropic.com>

sjg20 commented:
This has now been pending for over a year, so probably best to close it.
